core/src/main/scala/org/apache/spark/deploy/master/MasterSource.scala

```diff
@@ -42,6 +42,6 @@ private[spark] class MasterSource(val master: Master) extends Source {
 
   // Gauge for waiting application numbers in cluster
   metricRegistry.register(MetricRegistry.name("waitingApps"), new Gauge[Int] {
-    override def getValue: Int = master.waitingApps.size
+    override def getValue: Int = master.apps.filter(_.state == ApplicationState.WAITING).size
```

Comment:
Digging a bit about current usage of `waitingApps`, I think it cannot be well reflect the actual meaning of waiting applications, even RUNNING applications are still in the `waitingApps` list. So I think your fix is simple and effective.